### ISSUE TITLE
Complex content

### ISSUE THREAD

**Issue by Lewiscowles1986** [WordPress/gutenberg#150]
I Have a rather simple plugin https://github.com/CODESIGN2/WordPress-Content-Sections. I generally use on all client sites and even themes and child themes. 

 * It means people can update their site and easily manage content without
    * becoming programmers
    * losing that content during a theme change

I Even create plugins for that plugin. 

One thing I'm super interested in (tried with Visual Shortcode API but it was buggy) is being able to support complex content embedding (ideally both read-only reference (oEmbed?) and read-write template).

Read-write templates are basically things like two and three column blocks, and maybe suits widgets more.

Read only reference pulls in a piece of content so that it's easier to

 * edit non-widget rich content
    * without a pile of plugins
    * using WP editor
 * centrally manage, share and create content avoiding copy-pasta
 * augment third-party offerings

I'd love to see more ways to manage complex content like CPT's that are not meant to stand-alone, but rather be composed (in a similar vein to your ordering of elements.

**Comment by aduth** [comment]
Related: #104

The read-only references sound like a unique idea. Do you think the advantage of those is primarily in being able to reuse the content, or simply that it's previously been easier previously to edit the individual pieces of content given the nature of the editor?

**Comment by Lewiscowles1986** [comment]
Basically the read-only references prevent editing where referenced like an oEmbed of a SoundCloud or Gist or YouTube; they don't give rights to modify, just view. You'd only be able to change the order using Gutenberg. This would allow theoretical modifications and filters on `the_content` for URL parameters. 

Most of the benefit of chunks of content is you can experiment with content, ensure consistency and save time & money if you don't just put things up hoping for the best. It's like having a style guide where specific implementations are available through content-sections.

* The slight "aside" from how Gutenberg is organising that content is that it's central to the site, not the page. 
* The in-page reference is controlled in-page like all other Gutenberg elements, but not the content itself
* To allow editing the content, you'd probably only need an anchor or button and target="_blank" semantics
    * when done editing the user can close the screen and continue on the page they were on without losing changes
    * probably would need to trigger auto-save before navigating too, due to likelihood of some browsers or devices GC'ing inactive tabs (needs evidence).

I've actually had a brain-wave that might make Gutenberg harder for some content, but my life easier. I was thinking "How do I express this?". Turns out it's a lot like oEmbed, just using the host site as a source; so why don't I use oEmbed (I know silly me) for read-only references. We're currently doing that for a brand so their partners can access approved marketing copy and elements, so why not make an altered plugin to do the same for the current site...

Currently if you made a content-section called "contact-form" you could embed that in a contact page using shortcodes like so

```
[content_section name="contact-form"]
```

It's not very pretty but the plugin works. It also registers an Ajax endpoint that basically calls the shortcode and returns the output Html. This is what we cook up a minimal plugin for channel partners so they can link to the content, set caching rules in the webserver.

```
http://somesite.com/wp-admin/admin-ajax.php?action=get_content_section&name=some-content-slug
```

I'll link to this in my repo, and circle back if we find any issues; it's at the least helped me get more ideas :smile:

**Comment by aduth** [comment]
@Lewiscowles1986 Perhaps also related (or at least sounds similar to me) is @iseulde's exploration of an idea for resource identifiers:

https://make.wordpress.org/core/2016/12/24/idea-uniform-resource-identifiers-as-an-alternative-to-shortcodes/

**Comment by jasmussen** [comment]
Can we extract some actionable actions for Gutenberg from this ticket? Or is it more of a question, and if so, was it answered?

**Comment by jasmussen** [comment]
Closing for now, we can always reopen/revisit later.
